src/state/safety.rs

```diff
@@ -0,0 +1,315 @@
+///
+/// For this, it is necessary that:
+/// 1. Its chain ID matches the chain ID of the replica, or is the genesis qc.
+/// 2. It justifies a known block, or is the genesis qc.
+/// 3. Its view number is greater than the locked qc view or its block extends from the locked block.
+/// 4. If it is a prepare, precommit, or commit qc, the block it justifies has pending validator state
+///    updates.
+/// 5. If its qc is a generic qc, the block it justifies *does not* have pending validator set updates.
+///
+/// # Precondition
+/// [QuorumCertificate::is_correct] holds for block.justify.
+pub fn safe_qc<K: KVStore>(qc: &QuorumCertificate, block_tree: &BlockTree<K>, chain_id: ChainID) -> Result<bool, BlockTreeError> {
+    Ok(
+        /* 1 */ (qc.chain_id == chain_id || qc.is_genesis_qc()) &&
+        /* 2 */ (block_tree.contains(&qc.block) || qc.is_genesis_qc()) &&
+        /* 3 */ (qc.view > block_tree.locked_qc()?.view || extends_locked_qc_block(qc, block_tree)?) &&
+        /* 4 */ (((qc.phase.is_prepare() || qc.phase.is_precommit() || qc.phase.is_commit() || qc.phase.is_decide()) && 
+                   block_tree.validator_set_updates_status(&qc.block)?.contains_updates()) ||
+        /* 5 */ (qc.phase.is_generic() && !block_tree.validator_set_updates_status(&qc.block)?.contains_updates()))
+    )
+}
+
+/// Returns whether a nudge can be considered safe, and hence cause updates to the block tree.
+/// For this, it is necessary that:
+/// 1. safe_qc(&nudge.justify, block_tree, chain_id).
+/// 2. nudge.justify is a prepare, precommit, or commit qc.
+/// 3. Its chain ID matches the chain ID of the replica.
+/// 4. &nudge.justify is either a commit qc, or &nudge.justify.view = cur_view - 1.
+/// 
+/// This method enforces the commit rule for validator-set-updating blocks.
+/// 
+/// # Precondition
+/// [QuorumCertificate::is_correct] holds for nudge.justify.
+pub fn safe_nudge<K: KVStore>(nudge: &Nudge, cur_view: ViewNumber, block_tree: &BlockTree<K>, chain_id: ChainID) -> Result<bool, BlockTreeError> {
+    Ok(
+        /* 1 */ safe_qc(&nudge.justify, block_tree, chain_id)? &&
+        /* 2 */ nudge.justify.is_nudge_justify() &&
+        /* 3 */ (nudge.chain_id == chain_id) &&
+        /* 4 */ (nudge.justify.phase.is_commit() || nudge.justify.view == cur_view - 1)
+    )
+}
+
+/// Returns an optional qc that should be set as the new locked qc after seeing a given justify. This
+/// should be called whenever a new, correct and safe justify qc is seen, whether through a block
+/// proposal or through a nudge. This method implements the following rule:
+/// 1. If a block justify is seen,
+///     - if it is a decide qc, then locked qc should be updated to the block justify.
+///     - else the locked qc should be updated to the parent qc of the block justify.
+/// 2. If a nudge justify is seen,
+///     - if it is a precommit qc, then locked qc should be updated to the precommit qc.
+///     - else the locked qc should not be updated.
```

Comment:
So Locked QC is not updated when we receive a PrepareQC or a CommitQC, correct?

Why?

Reply:
I thought about this for a couple hours and couldn't think of any threats to safety from this. @karolinagrzeszkiewicz any thoughts from you are welcome, otherwise if there aren't any comments on this conversation I will resolve it tomorrow night.

Reply:
Thanks for bringing this up! It should be updated on seeing a prepareQC - for consistency with HotStuff, and because otherwise we might have some safety violations (if some replicas receive a commitQC for the block that should be locked but isn't and commit it, but others don't, and not being locked on the block they can vote for conflicting blocks). It's fixed now.

As for locking on seeing commitQC and decideQC in general there should be no need to do it, but in some circumstances we may need to do it - when the replica did not see a precommitQC for the given block.

You can find the updated version, with updated documentation explaining the implementation, in my [latest commit](https://github.com/parallelchain-io/hotstuff_rs/pull/36/commits/b36485bccc6aed22200fe6027c0b740ce03d242f).